src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -238,14 +238,20 @@ public uint GetConnectionId(SNIHandle handle, ref Guid clientConnectionId)
         /// <returns>SNI error status</returns>
         public uint WritePacket(SNIHandle handle, SNIPacket packet, bool sync)
         {
+            SNIPacket clonedPacket = packet.Clone();
+            uint result;
             if (sync)
             {
-                return handle.Send(packet.Clone());
+                result = handle.Send(clonedPacket);
+                clonedPacket.Dispose();
             }
             else
             {
-                return handle.SendAsync(packet.Clone());
+                clonedPacket.DisposeAfterWriteAsync = true;
```

Comment:
Instead of maintaining another state inside SNIPacket, wouldn't it be better to tell the SNIHandle to Dispose the packet after SendAsync completes? You could put an optional parameter in handle.SendAsync(SNIPacket packet, bool disposePacket) and it can be set to true from this call. This way the packet doesn't need to maintain a state.

Reply:
@saurabh500 
I am not very clear about what you meant.
https://github.com/dotnet/corefx/blob/68cfddeac516d7552124ca5028275e884bccbeb5/src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs#L527-L535
[Line#532](https://github.com/dotnet/corefx/blob/68cfddeac516d7552124ca5028275e884bccbeb5/src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs#L532) runs asynchronously, so I cannot put `packet.Dispose()` after the line. 
`Dispose()` should happen in the async context, not the main context.
I could not find anywhere for `Dispose()` other than inside of `SNIPacket.WriteToStreamAsync()`

Reply:
I was saying, pass the disposePacket=true from SNIHandle.SendAsync -> packet.writeToStreamAsync() so that that writeToStreamAsync can dispose the packet in continuation when this parameter is set to true.